scapy/contrib/knx.py

```diff
@@ -0,0 +1,620 @@
+        ByteField("reserved", None),
+        PacketField("control_endpoint", HPAI(), HPAI)
+    ]
+
+
+# KNX Standard v2.1 - 03_08_02 p34
+class KNXDisconnectResponse(Packet):
+    name = "DISCONNECT_RESPONSE"
+    fields_desc = [
+        ByteField("communication_channel_id", None),
+        ByteField("status", 0x00)
+    ]
+
+
+# KNX Standard v2.1 - 03_08_03 p22
+class KNXConfigurationRequest(Packet):
+    name = "CONFIGURATION_REQUEST"
+    fields_desc = [
+        ByteField("structure_length", 0x04),
+        ByteField("communication_channel_id", 0x01),
+        ByteField("sequence_counter", None),
+        ByteField("reserved", None),
+        PacketField("cemi", CEMI(), CEMI)
+    ]
+
+    def post_build(self, p, pay):
+        p = (len(p[:4])).to_bytes(1, byteorder='big') + p[1:]
+        return p + pay
+
+
+# KNX Standard v2.1 - 03_08_03 p22
+class KNXConfigurationACK(Packet):
+    name = "CONFIGURATION_ACK"
+    fields_desc = [
+        ByteField("structure_length", None),
+        ByteField("communication_channel_id", 0x01),
+        ByteField("sequence_counter", None),
+        ByteField("status", None)
+    ]
+
+    def post_build(self, p, pay):
+        p = (len(p)).to_bytes(1, byteorder='big') + p[1:]
+        return p + pay
+
+
+# KNX Standard v2.1 - 03_08_04 p.17
+class KNXTunnelingRequest(Packet):
+    name = "TUNNELING_REQUEST"
+    fields_desc = [
+        ByteField("structure_length", 0x04),
+        ByteField("communication_channel_id", 0x01),
+        ByteField("sequence_counter", None),
+        ByteField("reserved", None),
+        PacketField("cemi", CEMI(), CEMI)
+    ]
+
+    def post_build(self, p, pay):
+        p = (len(p[:4])).to_bytes(1, byteorder='big') + p[1:]
+        return p + pay
+
+
+# KNX Standard v2.1 - 03_08_04 p.18
+class KNXTunnelingACK(Packet):
+    name = "TUNNELING_ACK"
+    fields_desc = [
+        ByteField("structure_length", None),
+        ByteField("communication_channel_id", 0x01),
+        ByteField("sequence_counter", None),
+        ByteField("status", None)
+    ]
+
+    def post_build(self, p, pay):
+        p = (len(p)).to_bytes(1, byteorder='big') + p[1:]
+        return p + pay
+
+
+### KNX FRAME
+
+# we made the choice to define a KNX service as a payload for a KNX Header
+# it could also be possible to define the body as a conditionnal PacketField contained after the header
+
+class KNX(Packet):
+    name = "KNXnet/IP"
+    fields_desc = [
+        ByteField("header_length", None),
+        XByteField("protocol_version", 0x10),
+        ShortEnumField("service_identifier", None, SERVICE_IDENTIFIER_CODES),
+        ShortField("total_length", None)
+    ]
+
+    def post_build(self, p, pay):
+        # computes header_length
+        p = (len(p)).to_bytes(1, byteorder='big') + p[1:]
+        # computes total_length
+        p = p[:-2] + (len(p) + len(pay)).to_bytes(2, byteorder='big')
+        return p + pay
+
+
+### LAYERS BINDING
+bind_layers(UDP, KNX, dport=3671)
```

Comment:
You would do
```
bind_bottom_up(UDP, KNW, dport=3671)
bind_bottom_up(UDP, KNW, sport=3671)
bind_layers(UDP, KNW, sport=3671, dport=3671)
```
to avoid any conflicts.